src/python/pants/core/util_rules/environments.py

```diff
@@ -146,54 +148,82 @@ class AmbiguousEnvironmentError(Exception):
     pass
 
 
+class UnrecognizedEnvironmentError(Exception):
+    pass
+
+
 class AllEnvironmentTargets(FrozenDict[str, LocalEnvironmentTarget]):
     """A mapping of environment aliases to their corresponding environment target."""
 
 
 @dataclass(frozen=True)
-class ChosenLocalEnvironment:
-    tgt: LocalEnvironmentTarget | None
+class ChosenLocalEnvironmentAlias:
+    f"""Which environment alias from `[environments-preview].aliases` that
+    {LOCAL_ENVIRONMENT_MATCHER} resolves to."""
+
+    val: str | None
+
+
+@dataclass(frozen=True)
+class ResolvedEnvironmentAlias(EngineAwareParameter):
+    f"""The normalized alias for an environment, from `[environments-preview].aliases`, after
+    applying things like {LOCAL_ENVIRONMENT_MATCHER}.
+
+    Note that we have this type, rather than only `ResolvedEnvironmentTarget`, for a more efficient
+    rule graph. This node impacts the equality of many downstream nodes, so we want its identity
+    to only be a single string, rather than a Target instance.
```

Comment:
@stuhood is this right?

Reply:
Yea, that's right.

Relatedly though: this type will need to be very widely used in `@rule` APIs: for example, in #16682, I will need to add it to many/most `@union` declarations. So giving it a shorter name now (even if the other types in here stay verbose) would be good: `EnvironmentId`, `EnvironmentAlias`, `EnvironmentName`, etc.